File: server/test/weather.test.js
/* */

import { use, expect } from "chai";
import chaiHttp from "chai-http";
const chai = use(chaiHttp);

const url = "http://localhost:3000";

describe('WEATHER TEST ', function () {
  before(function (done) {
    done();
  });
  if (url !== 'http://localhost:3000') {
    it('local temp // fails on localhost', function (done) {
      chai.request.execute(url)
        .get('/v1/weather')
        .query({})
        .end(function (err, res) {
          // console.log('PATH=> ', res.req.path)
          expect(err).to.be.null;
          expect(res).to.have.status(200);
          expect(res).to.be.json;
          expect(res.body).to.have.property('city');
          expect(res.body).to.have.property('country');
          expect(res.body).to.have.property('tempC');
          expect(res.body).to.have.property('tempF');
          done();
        });
    });
  }
  if (url === 'http://localhost:3000') {
    it('empty city', function (done) {
      chai.request.execute(url)
        .get('/v1/weather?city=')
        .query({})
        .end(function (err, res) {
          // console.log('PATH=> ', res.req.path)
          expect(err).to.be.null;
          expect(res).to.have.status(500);
          done();
        });
    });
  } else {
    it('local temp', function (done) {
      chai.request.execute(url)
        .get('/weather')
        .query({})
        .end(function (err, res) {
          // console.log('PATH=> ', res.req.path)
          expect(err).to.be.null;
          expect(res).to.have.status(200);
          expect(res).to.be.json;
          expect(res.body).to.have.property('city');
          expect(res.body).to.have.property('country');
          expect(res.body).to.have.property('tempC');
          expect(res.body).to.have.property('tempF');
          done();
        });
    });
  }
  it('unavailable city', function (done) {
    const t2 = "Weather Service Error: HTTP Error: 400 Bad Request";
    chai.request.execute(url)
      .get('/v1/weather?city=unavailablecity')
      .query({})
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(500);
        expect(res).to.be.json;
        expect(res.body).to.have.property('msg', t2);
        done();
      });
  });
  it('london city ', function (done) {
    chai.request.execute(url)
      .get('/v1/weather?city=london')
      .query({})
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(200);
        expect(res).to.be.json;
        expect(res.body).to.have.property('city');
        expect(res.body).to.have.property('country');
        expect(res.body).to.have.property('tempC');
        expect(res.body).to.have.property('tempF');
        done();
      });
  });
  it('new york city', function (done) {
    chai.request.execute(url)
      .get('/v1/weather?city=new york')
      .query({})
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(200);
        expect(res).to.be.json;
        expect(res.body).to.have.property('city');
        expect(res.body).to.have.property('country');
        expect(res.body).to.have.property('tempC');
        expect(res.body).to.have.property('tempF');
        done();
      });
  });
});

File: codetabs.0.9.2/weather/weather_test.go
/* */

package weather

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestWeatherApi(t *testing.T) {

	const (
		validFormat = "Bad request, valid format is 'api.codetabs.com/v1/{service}?{param}=value' .Please read our docs at https://codetabs.com"
	)

	w := NewWeather(true)

	type weatherTestOutput struct {
		StatusCode int     // `json:"statusCode"`
		Error      string  `json:"Error,omitempty"`
		TempC      float64 `json:"tempC"`
		TempF      float64 `json:"tempF"`
		City       string  `json:"city"`
		Country    string  `json:"country"`
		Lat        float64 `json:"latitude"`
		Lon        float64 `json:"longitude"`
	}

	var weatherTests = []struct {
		it         string
		endpoint   string
		errorText  string
		statusCode int
	}{
		{"empty", "/v1/weather", "", 200},
		{"bad city", "/v1/weather?city=i-dont-exist", "", 200},
		{"bad city", "/v1/weather/?city=lon%%don", "", 200},
		{"incomplete", "/v1/weather/?city=&format=xml", "", 200},
		{"bad", "/v1/weather/bad", validFormat, 400},
		{"local temp", "/v1/weather/", "", 200},
		{"empty city", "/v1/weather/?city=", "", 200},
		{"london", "/v1/weather/?city=london", "", 200},
		{"new york", "/v1/weather?city=new york", "", 200},
		{"", "/v1/weather?format=json", "", 200},
		{"", "/v1/weather?format=invalid", validFormat, 400},
		{"", "/v1/weather?format=xml", "", 200},
		//{"", "/v1/weather", "", 00},
	}

	for _, test := range weatherTests {
		var to = weatherTestOutput{}
		pass := true
		//fmt.Println(`Test...`, test.endpoint, "...", test.it)
		req, err := http.NewRequest("GET", test.endpoint, nil)
		//fmt.Println(`------------------------------`)
		//fmt.Println(err.Error())
		//fmt.Println(test.errorText)
		//fmt.Println(`------------------------------`)
		if err != nil {
			if err.Error() != test.errorText {
				t.Fatalf("Error Request %s\n", err.Error())
			} else {
				pass = false
			}
		}
		if pass {
			rr := httptest.NewRecorder()
			handler := http.HandlerFunc(w.Router)
			handler.ServeHTTP(rr, req)
			if rr.Code != test.statusCode {
				t.Errorf("%s got %v want %v\n", test.endpoint, rr.Code, test.statusCode)
			}
			_ = json.Unmarshal(rr.Body.Bytes(), &to)
			if to.Error != test.errorText {
				t.Errorf("%s got %v want %v\n", test.endpoint, to.Error, test.errorText)
			}
		}
		fmt.Printf("Test OK...%s\n", test.endpoint)
	}
}
